pkg/util/active_groups.go

```diff
@@ -0,0 +1,149 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package util
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/grafana/dskit/services"
+	"go.uber.org/atomic"
+)
+
+type ActiveGroups struct {
+	mu                sync.RWMutex
+	timestampsPerUser map[string]map[string]*atomic.Int64 // map[user][group] -> timestamp
+	maxGroupsPerUser  int
+}
+
+func NewActiveGroups(maxGroupsPerUser int) *ActiveGroups {
+	return &ActiveGroups{
+		timestampsPerUser: map[string]map[string]*atomic.Int64{},
+		maxGroupsPerUser:  maxGroupsPerUser,
+	}
+}
+
+// UpdateGroupTimestampForUser function is only guaranteed to update to timestamp
+// provided even if it is smaller than the existing value
+func (ag *ActiveGroups) UpdateGroupTimestampForUser(userID, group string, ts int64) {
```

Comment:
Nit: I'd make ActiveGroups' methods accept `time.Time` instead of `int64`. The fact that it needs to store them as int64 is an implementation detail and shouldn't be exposed to outside (which always raises the question: is this a unix timestamp? is this unix millis?)